Split gas optimisation

rainbreak opened the pull request:
Addressing #31.

Change split process from split(i) -> j, k to split(i) -> i', j.
That is, modify the existing auctionlet rather than deleting it and creating a new
one.

Note that this gives an important change to auction semantics: previously,
the quantity being bid on in an auctionlet was invariant. Now, bidders cannot
be assured that a given auctionlet will maintain the bid quantity they know it
to have.

External agents must pay close attention to the auction event stream, or they
could end up successfully bidding for a much lower quantity than they expected.

Gas before:
    - base split: 448k
    - non base split: 418k

Gas after:
    - base split: 280k
    - non base split: 234k

rainbreak committed:
split optimise: modify bid, rather than replace

Change split process from split(i) -> j, k to split(i) -> i', j.
That is, modify the existing bid rather than deleting it and creating a new
one.

Note that this gives an important change to auction semantics: previously,
the quantity being bid on in an auctionlet was invariant. Now, bidders cannot
be assured that a given auctionlet will maintain the bid quantity they know it
to have.

External agents must pay close attention to the auction event stream, or they
could end up successfully bidding for a much lower quantity than they expected.

Gas before:
    - base split: 448k
    - non base split: 418k

Gas after:
    - base split: 296k
    - non base split: 250k

rainbreak committed:
split optimise: remove redundant code

Gas before:
    - base split: 296k
    - non base split: 250k

Gas after:
    - base split: 280k
    - non base split: 234k

rainbreak committed:
don't set last bid time in newAuctionlet

The last bid time has no bearing on base auctionlets (where the `base` flag
matters) and should really only be modified at bid time.

rainbeam commented:
@freekp @jorisbontje: any thoughts on the change to UX this would mean? Previously, an auctionlet could be taken to represent a known quantity of lot being sold. With this PR, bidders would have to be aware that the quantity associated with an auctionlet can change.

In practice, my worst imagined scenario looks like this (in a regular auction selling 20 widgets):

- Alice bids $100 for 20 widgets (on auctionlet id 1)
- Bob splits id 1, offering $30 for 5 widgets (creating auctionlet id 2)
- Resulting auctionlets:
    - 1: $75 for 15 widgets (high bidder is Alice)
    - 2: $30 for 5 widgets (high bidder is Bob)
- Charlie bids $120 on auctionlet 1. However, Charlie didn't pay attention to the `Split` event and thought he was bidding for 20 widgets, when he is actually bidding for 15.

There is a way out of this, which is to require all bids (not just split bids) to be of the form 

    bid(auctionlet, bid_how_much, quantity)

i.e. always require a specific quantity, rather than implicitly taking the quantity from the auctionlet in the case of `bid(auctionlet, bid_how_much)`.

jorisbontje commented:
An alternative (lesser) approach to adding the `quantity` field, would be to introduce an auctionlet `version` number which would need to be provided for each bid; to verify that the user is looking at the right version. This would require additional storage, and version tracking; while providing the `quantity` directly captures the intend. 

I am in favour of requiring `bid(auctionlet, bid_how_much, quantity)` consistently.

nmushegian commented:
Not the first place where a fill-or-kill pattern would be useful. I think you only need 2 or maybe even 1 byte for the ticket/counter.

rainbeam commented:
Alternatives as I see them:

1. Specify the *bid* to beat with a pair `(auctionlet_id, bid_id)`:

    ```
    bid(auctionlet_id, bid_id, bid_how_much)
    ```

    - fail if `bid_id != auctionlet.nbid`, bid otherwise
    - split bids need to specify quantity as well

2. Uniquely identify bids with `bid_id` (internally use mapping(uint=>uint) to link the id to the associated auctionlet)

    ```
    bid(bid_id, bid_how_much)
    ```

    - fail if `bid_id` not found (deleted with state change)
    - splits specify quantity as well

3. Specify the quantity wanted from a specific auctionlet:

    ```
    bid(auctionlet_id, bid_how_much, quantity)
    ```

    - fail if `quantity > auctionlet.sell_amount`, 
    - bid if `quantity == auctionlet.sell_amount`
    - split bid if `quantity < auctionlet.sell_amount`


Without this PR, bidding is on auctionlets, which are invariant w.r.t quantity. (1) and (2) move from bidding on auctionlets to bidding on invariant 'bidlets'. (3) makes auctionlets mutable, but the bidder is explicit about what they want.

rainbreak committed:
db: make getter public

rainbreak committed:
separate auction manager controller and frontend

This also allows for easier control of public functions by packages that use
this as a base. They can inherit from AuctionController and set custom initialisers.

rainbreak committed:
force quantity in split bid

freekp committed:
New deploy of contract

rainbreak committed:
update readme

rainbeam commented:
Merging this now. Have gone with approach 3. All tests passing.
